include/dlaf/eigensolver/eigensolver.h

```diff
@@ -52,9 +52,15 @@ namespace dlaf {
 /// @pre @p eigenvectors has size (N x N)
 /// @pre @p eigenvectors has blocksize (NB x NB)
 /// @pre @p eigenvectors has tilesize (NB x NB)
+///
+/// @param[in] first_eigenvalue_index is the index of the first eigenvalue to compute
+/// @pre @p first_eigenvalue_index == 0
+/// @param[in] last_eigenvalue_index is the index of the last eigenvalue to compute
+/// @pre @p last_eigenvalue_index < N
 template <Backend B, Device D, class T>
 void hermitian_eigensolver(blas::Uplo uplo, Matrix<T, D>& mat, Matrix<BaseType<T>, D>& eigenvalues,
-                           Matrix<T, D>& eigenvectors) {
+                           Matrix<T, D>& eigenvectors, SizeType first_eigenvalue_index,
+                           SizeType last_eigenvalue_index) {
```

Comment:
> However, I'm more favorable to [first, last), i.e last excluded.

ScaLAPACK uses `[first, last]`, do we want to differ? However it is Fortran indexing so it is less ugly (i.e. the last index can be `N`, not `N-1`).

Reply:
I don't have a strong opinion on this, but from the options so far it looks to me like [first, last) seems like the sensible option because:
- it's still two indices like scalapack, but
- it's still the more natural formulation for 0-based indices

I think the difference to scalapack's indexing isn't too bad, but I'm not the typical user... Do we create massive confusion if the C API uses [first, last) with 0-based indices and the Fortran API [first, last] with 1-based indices?

Reply:
We discussed this yesterday at the meeting, and the consensus was essentially what you suggest here. Using `[first, last)` is much cleaner in C/C++, and the Fortran API can still expose `[first, last]`.

> Do we create massive confusion if the C API uses [first, last) with 0-based indices and the Fortran API [first, last] with 1-based indices?

I don't think so. I think differentiating between C/C++ and Fortran will make it less confusing for the different users.

https://github.com/eth-cscs/DLA-Future/pull/1194/commits/f566f8980cbbbf326222daa7688eab11d56bb826 takes care of most of the changes already; I also changed `first`/`last` with `begin`/`end`.